### DISCUSSION TITLE
LDM Notes for October 17th, 2022

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#6569] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-10-17.md

## Agenda

- Primary Constructors
- Permit variable declarations under disjunctive patterns

**Comment by jnm2** [comment]
`init { }` makes me think of final validators (which would run as the last part of all the `init` things) and it makes me wonder how they'd look alongside each other.

**Reply by MadsTorgersen** [reply]
This is a good point to call out. Just last week I drafted a [proposal](https://github.com/dotnet/csharplang/blob/construction-improvements/proposals/final-initializers.md) for final initializers that uses that syntax.

**Reply by sab39** [reply]
For consideration: to what extent, if at all, would final initializers by themselves meet the needs that motivate the desire for primary constructor bodies?

**Reply by MadsTorgersen** [reply]
It's worth a thought, and I did mention it in the proposal. I think the main thing that they wouldn't be ideal for is argument validation, which folks tend to prefer to do up front before any other work is done.

There's definitely a risk that it will be confusing to have both and have to reason about where everything should go.

**Reply by qrli** [reply]
`ClassName { }` could be too verbose, especially when the class name is long, which defeats the purpose of primary constructor.
If `init { }` is to be used for final initializer, why not use `new { }` for primary constructor body.

**Reply by Joe4evr** [reply]
@MadsTorgersen Is there a proper discussion thread for Final initializers yet?

**Reply by HaloFour** [reply]
With autocomplete `ClassName { }` would require probably the same number of keystrokes as `new { }` or `init { }`.  It's been stated that if C# were to be redesigned that the constructor syntax (if constructors were to still exist) might be reconsidered, but I think it's better to remain consistent with the language we have today.

Java uses `ClassName { }` for the constructor body with records:

```java
public record Person(String name) {

    Person {
        Objects.requireNotNull(name, "name is null");
    }
}
```

**Reply by qrli** [reply]
Putting record aside and talking about primary constructor only, compare
```cs
public class Person(String name) {
    Person {
        Objects.requireNotNull(name, "name is null");
    }
}
```
to
```cs
public class Person {
    Person(String name) {
        Objects.requireNotNull(name, "name is null");
    }
}
```
Why bother?
And I'd argue primary constructor _body_ is not the same thing as _constructor_.

**Reply by HaloFour** [reply]
> Why bother?

Scope.  You can't access `name` outside of the constructor body in the second example.  My opinion is that this will be useful primarily for components that use dependency injection so that you can skip the boilerplate of field definitions and assignments for those dependencies.

**Comment by orthoxerox** [comment]
The biggest pain point for record constructors that I would love to see tackled by primary constructors is the need to switch to a regular constructor as soon as you want one of the following:

- validation (the biggest issue). `!!` would've solved like 80% of the cases, but not all of them
- different types on properties. `params Foo[]` to `IReadOnlyList<Foo>` is, again, 80% of the cases
